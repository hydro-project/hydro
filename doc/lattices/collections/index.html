<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Simple singleton or array collection with `cc_traits` implementations."><title>lattices::collections - Rust</title><script>if(window.location.protocol!=="file:")document.head.insertAdjacentHTML("beforeend","SourceSerif4-Regular-6b053e98.ttf.woff2,FiraSans-Italic-81dc35de.woff2,FiraSans-Regular-0fe48ade.woff2,FiraSans-MediumItalic-ccf7e434.woff2,FiraSans-Medium-e1aa3f0a.woff2,SourceCodePro-Regular-8badfe75.ttf.woff2,SourceCodePro-Semibold-aa29a496.ttf.woff2".split(",").map(f=>`<link rel="preload" as="font" type="font/woff2" crossorigin href="../../static.files/${f}">`).join(""))</script><link rel="stylesheet" href="../../static.files/normalize-9960930a.css"><link rel="stylesheet" href="../../static.files/rustdoc-20739d33.css"><meta name="rustdoc-vars" data-root-path="../../" data-static-root-path="../../static.files/" data-current-crate="lattices" data-themes="" data-resource-suffix="" data-rustdoc-version="1.88.0-nightly (10fa3c449 2025-04-26)" data-channel="nightly" data-search-js="search-f7877310.js" data-settings-js="settings-5514c975.js" ><script src="../../static.files/storage-4e99c027.js"></script><script defer src="../sidebar-items.js"></script><script defer src="../../static.files/main-7ef8a74a.js"></script><noscript><link rel="stylesheet" href="../../static.files/noscript-893ab5e7.css"></noscript><link rel="alternate icon" type="image/png" href="../../static.files/favicon-32x32-6580c154.png"><link rel="icon" type="image/svg+xml" href="../../static.files/favicon-044be391.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle" title="show sidebar"></button></nav><nav class="sidebar"><div class="sidebar-crate"><h2><a href="../../lattices/index.html">lattices</a><span class="version">0.6.1</span></h2></div><div class="sidebar-elems"><section id="rustdoc-toc"><h2 class="location"><a href="#">Module collections</a></h2><h3><a href="#structs">Module Items</a></h3><ul class="block"><li><a href="#structs" title="Structs">Structs</a></li><li><a href="#traits" title="Traits">Traits</a></li></ul></section><div id="rustdoc-modnav"><h2 class="in-crate"><a href="../index.html">In crate lattices</a></h2></div></div></nav><div class="sidebar-resizer"></div><main><div class="width-limiter"><rustdoc-search></rustdoc-search><section id="main-content" class="content"><div class="main-heading"><div class="rustdoc-breadcrumbs"><a href="../index.html">lattices</a></div><h1>Module <span>collections</span><button id="copy-path" title="Copy item path to clipboard">Copy item path</button></h1><rustdoc-toolbar></rustdoc-toolbar><span class="sub-heading"><a class="src" href="../../src/lattices/collections.rs.html#1-1203">Source</a> </span></div><details class="toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Simple singleton or array collection with [<code>cc_traits</code>] implementations.</p>
</div></details><h2 id="structs" class="section-header">Structs<a href="#structs" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="struct" href="struct.ArrayMap.html" title="struct lattices::collections::ArrayMap">Array<wbr>Map</a></dt><dd>An array wrapper representing a fixed-size map.</dd><dt><a class="struct" href="struct.ArraySet.html" title="struct lattices::collections::ArraySet">Array<wbr>Set</a></dt><dd>An array wrapper representing a fixed-size set (modulo duplicate items).</dd><dt><a class="struct" href="struct.EmptyMap.html" title="struct lattices::collections::EmptyMap">Empty<wbr>Map</a></dt><dd>A key-value entry wrapper representing a singleton map.</dd><dt><a class="struct" href="struct.EmptySet.html" title="struct lattices::collections::EmptySet">Empty<wbr>Set</a></dt><dd>A type that will always be an empty set.</dd><dt><a class="struct" href="struct.OptionMap.html" title="struct lattices::collections::OptionMap">Option<wbr>Map</a></dt><dd>A key-value entry wrapper around <code>Option&lt;(K, V)&gt;</code> representing a singleton or empty map.</dd><dt><a class="struct" href="struct.OptionSet.html" title="struct lattices::collections::OptionSet">Option<wbr>Set</a></dt><dd>A wrapper around <code>Option</code>, representing either a singleton or empty set.</dd><dt><a class="struct" href="struct.SingletonMap.html" title="struct lattices::collections::SingletonMap">Singleton<wbr>Map</a></dt><dd>A key-value entry wrapper representing a singleton map.</dd><dt><a class="struct" href="struct.SingletonSet.html" title="struct lattices::collections::SingletonSet">Singleton<wbr>Set</a></dt><dd>A wrapper around an item, representing a singleton set.</dd><dt><a class="struct" href="struct.VecMap.html" title="struct lattices::collections::VecMap">VecMap</a></dt><dd>A <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>-wrapper representing a naively implemented map.</dd><dt><a class="struct" href="struct.VecSet.html" title="struct lattices::collections::VecSet">VecSet</a></dt><dd>A <a href="https://doc.rust-lang.org/nightly/alloc/vec/struct.Vec.html" title="struct alloc::vec::Vec"><code>Vec</code></a>-wrapper representing a naively-implemented set.</dd></dl><h2 id="traits" class="section-header">Traits<a href="#traits" class="anchor">ยง</a></h2><dl class="item-table"><dt><a class="trait" href="trait.MapMapValues.html" title="trait lattices::collections::MapMapValues">MapMap<wbr>Values</a></dt><dd>Trait for transforming the values of a map without changing the overall type of the data structure.</dd></dl></section></div></main></body></html>