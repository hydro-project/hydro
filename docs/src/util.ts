export function getLines(str: string, lineStart: number, lineEnd?: number): string {
    return str.split('\n').slice(lineStart - 1, lineEnd || lineStart).join('\n');
}

export function extractOutput(output: string): string {
    const outputLines = output.replace(/\n$/, '').split('\n');
    outputLines.splice(0, 4);
    if (outputLines[0].startsWith('%%')) {
        const count = outputLines.findIndex(line => 0 === line.length);
        outputLines.splice(0, count + 1, '<graph output>');
    }
    const stdOut = outputLines.join('\n');
    return `#shell-command-next-line
cargo run
<build output>
${stdOut}`;
}

export function extractMermaid(output: string): string {
    const outputLines = output.split('\n');
    outputLines.splice(0, 4);
    const count = outputLines.findIndex(line => 0 === line.length);
    outputLines.length = count;
    return outputLines.join('\n');
}