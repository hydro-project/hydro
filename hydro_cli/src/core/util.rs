use std::time::Duration;

use anyhow::Result;
use futures::Future;

pub async fn async_retry<T, F: Future<Output = Result<T>>>(
    thunk: impl Fn() -> F,
    count: usize,
    delay: Duration,
) -> Result<T> {
    for _ in 1..count {
        let result = thunk().await;
        if result.is_ok() {
            return result;
        } else {
            tokio::time::sleep(delay).await;
        }
    }

    thunk().await
}
