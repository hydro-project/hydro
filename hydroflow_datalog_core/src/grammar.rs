#[rust_sitter::grammar("datalog")]
#[allow(dead_code)]
pub mod datalog {
    #[rust_sitter::language]
    #[derive(Debug)]
    pub struct Program {
        pub rules: Vec<Declaration>,
    }

    #[derive(Debug, Clone)]
    pub enum Declaration {
        Input(#[rust_sitter::leaf(text = ".input")] (), Ident, RustSnippet),
        Output(
            #[rust_sitter::leaf(text = ".output")] (),
            Ident,
            RustSnippet,
        ),
        Async(
            #[rust_sitter::leaf(text = ".async")] (),
            Ident,
            /// A pipeline for data to be sent to another node, which must consume `(NodeID, Data)` pairs.
            RustSnippet,
            /// A pipeline for data received from another node, which must produce `Data` values.
            RustSnippet,
        ),
        Rule(Rule),
    }

    #[derive(Debug, Clone)]
    pub struct RustSnippet {
        #[rust_sitter::leaf(text = "`")]
        _start: (),
        #[rust_sitter::leaf(pattern = r#"[^`]*"#, transform = |s| s.to_string())]
        pub code: String,
        #[rust_sitter::leaf(text = "`")]
        _end: (),
    }

    #[derive(Debug, Clone)]
    pub struct Rule {
        pub target: TargetRelationExpr,

        pub rule_type: RuleType,

        #[rust_sitter::repeat(non_empty = true)]
        #[rust_sitter::delimited(
            #[rust_sitter::leaf(text = ",")]
            ()
        )]
        pub sources: Vec<Atom>,

        #[rust_sitter::leaf(text = ".")]
        _dot: Option<()>,
    }

    #[derive(Debug, Clone)]
    pub enum RuleType {
        Sync(#[rust_sitter::leaf(text = ":-")] ()),
        NextTick(#[rust_sitter::leaf(text = ":+")] ()),
        Async(#[rust_sitter::leaf(text = ":~")] ()),
    }

    #[derive(Debug, Clone)]
    pub enum Atom {
        Relation(
            #[rust_sitter::leaf(text = "!")] Option<()>,
            InputRelationExpr,
        ),
        Predicate(PredicateExpr),
    }

    #[derive(Debug, Clone)]
    pub struct InputRelationExpr {
        pub name: Ident,

        #[rust_sitter::leaf(text = "(")]
        _l_paren: (),

        #[rust_sitter::delimited(
            #[rust_sitter::leaf(text = ",")]
            ()
        )]
        pub fields: Vec<Ident>,

        #[rust_sitter::leaf(text = ")")]
        _r_paren: (),
    }

    #[derive(Debug, Clone)]
    pub struct TargetRelationExpr {
        pub name: Ident,

        pub at_node: Option<AtNode>,

        #[rust_sitter::leaf(text = "(")]
        _l_paren: (),

        #[rust_sitter::delimited(
            #[rust_sitter::leaf(text = ",")]
            ()
        )]
        pub fields: Vec<TargetExpr>,

        #[rust_sitter::leaf(text = ")")]
        _r_paren: (),
    }

    #[derive(Debug, Clone)]
    #[allow(clippy::manual_non_exhaustive)]
    pub struct AtNode {
        #[rust_sitter::leaf(text = "@")]
        _at: (),

        pub node: TargetExpr,
    }

    #[derive(Debug, Clone)]
    pub enum TargetExpr {
        Ident(Ident),
        Aggregation(Aggregation),
    }

    impl TargetExpr {
        pub fn ident(&self) -> &Ident {
            match self {
                TargetExpr::Ident(ident) => ident,
                TargetExpr::Aggregation(a) => &a.ident,
            }
        }
    }

    #[derive(Debug, Clone)]
    pub struct Aggregation {
        pub tpe: AggregationType,
        #[rust_sitter::leaf(text = "(")]
        _lparen: (),
        pub ident: Ident,
        #[rust_sitter::leaf(text = ")")]
        _rparen: (),
    }

    #[derive(Debug, Clone)]
    pub enum AggregationType {
        Max(#[rust_sitter::leaf(text = "max")] ()),
    }

    #[derive(Clone, PartialEq, Eq, Hash, Ord, PartialOrd, Debug)]
    pub struct Ident {
        #[rust_sitter::leaf(pattern = r"[a-zA-Z_][a-zA-Z0-9_]*", transform = |s| s.to_string())]
        pub name: String,
    }

    #[rust_sitter::extra]
    struct Whitespace {
        #[rust_sitter::leaf(pattern = r"\s")]
        _whitespace: (),
    }

    #[derive(Debug, Clone)]
    pub enum BoolOp {
        Lt(#[rust_sitter::leaf(text = "<")] ()),
        LtEq(#[rust_sitter::leaf(text = "<=")] ()),
        Gt(#[rust_sitter::leaf(text = ">")] ()),
        GtEq(#[rust_sitter::leaf(text = ">=")] ()),
        Eq(#[rust_sitter::leaf(text = "==")] ()),
    }

    #[derive(Debug, Clone)]
    pub struct PredicateExpr {
        #[rust_sitter::leaf(text = "(")]
        _l_brace: (),

        pub left: Ident,
        pub op: BoolOp,
        pub right: Ident,

        #[rust_sitter::leaf(text = ")")]
        _r_brace: (),
    }
}
