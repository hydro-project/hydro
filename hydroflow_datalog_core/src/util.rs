use std::ops::RangeFrom;

pub type Counter = RangeFrom<isize>;
