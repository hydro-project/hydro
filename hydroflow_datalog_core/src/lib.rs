use std::collections::{HashMap, HashSet};

use hydroflow_lang::{
    diagnostic::{Diagnostic, Level},
    graph::{flat_to_partitioned::partition_graph, FlatGraphBuilder, PartitionedGraph},
    parse::{IndexInt, Indexing, Pipeline, PipelineLink},
};
use proc_macro2::{Span, TokenStream};
use syn::parse_quote;

mod grammar;
mod join_plan;
mod util;

use grammar::datalog::*;
use join_plan::*;
use util::{repeat_tuple, Counter};

pub fn gen_hydroflow_graph(
    literal: proc_macro2::Literal,
) -> Result<PartitionedGraph, (Vec<rust_sitter::errors::ParseError>, Vec<Diagnostic>)> {
    let str_node: syn::LitStr = parse_quote!(#literal);
    let actual_str = str_node.value();
    let program: Program = grammar::datalog::parse(&actual_str).map_err(|e| (e, vec![]))?;

    let mut inputs = Vec::new();
    let mut outputs = Vec::new();
    let mut asyncs = Vec::new();
    let mut rules = Vec::new();

    for stmt in &program.rules {
        match stmt {
            Declaration::Input(_, ident, hf_code) => inputs.push((ident, hf_code)),
            Declaration::Output(_, ident, hf_code) => outputs.push((ident, hf_code)),
            Declaration::Async(_, ident, send_hf, recv_hf) => {
                asyncs.push((ident, send_hf, recv_hf))
            }
            Declaration::Rule(rule) => rules.push(rule),
        }
    }

    let mut flat_graph_builder = FlatGraphBuilder::new();
    let mut tee_counter = HashMap::new();
    let mut merge_counter = HashMap::new();

    let mut created_rules = HashSet::new();
    for decl in &program.rules {
        let target_ident = match decl {
            Declaration::Input(_, ident, _) => ident.clone(),
            Declaration::Output(_, ident, _) => ident.clone(),
            Declaration::Async(_, ident, _, _) => ident.clone(),
            Declaration::Rule(rule) => rule.target.name.clone(),
        };

        if !created_rules.contains(&target_ident) {
            created_rules.insert(target_ident.clone());
            let name = syn::Ident::new(&target_ident.name, Span::call_site());
            flat_graph_builder
                .add_statement(parse_quote!(#name = merge() -> unique::<'tick>() -> tee()));
        }
    }

    for (target, hf_code) in inputs {
        let my_merge_index = merge_counter
            .entry(target.name.clone())
            .or_insert_with(|| 0..)
            .next()
            .expect("Out of merge indices");

        let my_merge_index_lit =
            syn::LitInt::new(&format!("{}", my_merge_index), Span::call_site());
        let name = syn::Ident::new(&target.name, Span::call_site());

        let input_pipeline: Pipeline = syn::parse_str(&hf_code.code).unwrap();

        flat_graph_builder.add_statement(parse_quote! {
            #input_pipeline -> [#my_merge_index_lit] #name
        });
    }

    for (target, hf_code) in outputs {
        let my_tee_index = tee_counter
            .entry(target.name.clone())
            .or_insert_with(|| 0..)
            .next()
            .expect("Out of tee indices");

        let my_tee_index_lit = syn::LitInt::new(&format!("{}", my_tee_index), Span::call_site());
        let target_ident = syn::Ident::new(&target.name, Span::call_site());

        let output_pipeline: Pipeline = syn::parse_str(&hf_code.code).unwrap();

        flat_graph_builder.add_statement(parse_quote! {
            #target_ident [#my_tee_index_lit] -> #output_pipeline
        });
    }

    for (target, send_hf, recv_hf) in asyncs {
        let async_send_pipeline = format!("{}_async_send", target.name);
        let async_send_pipeline = syn::Ident::new(&async_send_pipeline, Span::call_site());

        let recv_merge_index = merge_counter
            .entry(target.name.clone())
            .or_insert_with(|| 0..)
            .next()
            .expect("Out of merge indices");

        let recv_merge_index_lit =
            syn::LitInt::new(&format!("{}", recv_merge_index), Span::call_site());
        let target_ident = syn::Ident::new(&target.name, Span::call_site());

        let send_pipeline: Pipeline = syn::parse_str(&send_hf.code).unwrap();
        let recv_pipeline: Pipeline = syn::parse_str(&recv_hf.code).unwrap();

        flat_graph_builder.add_statement(parse_quote! {
            #async_send_pipeline = merge() -> #send_pipeline
        });

        flat_graph_builder.add_statement(parse_quote! {
            #recv_pipeline -> [#recv_merge_index_lit] #target_ident
        });
    }

    let mut next_join_idx = 0..;
    let mut diagnostics = Vec::new();
    for rule in rules {
        generate_rule(
            rule,
            &mut flat_graph_builder,
            &mut tee_counter,
            &mut merge_counter,
            &mut next_join_idx,
            &mut diagnostics,
        );
    }

    if !diagnostics.is_empty() {
        Err((vec![], diagnostics))
    } else {
        let flat_graph = flat_graph_builder
            .build(Level::Error)
            .unwrap_or_else(std::convert::identity);
        Ok(flat_graph)
    }
}

pub fn hydroflow_graph_to_program(flat_graph: PartitionedGraph, root: TokenStream) -> syn::Stmt {
    let partitioned_graph =
        partition_graph(flat_graph).expect("Failed to partition (cycle detected).");
    let code_tokens = partitioned_graph.as_code(root, true);

    syn::parse_quote!({
        #code_tokens
    })
}

fn generate_rule(
    rule: &Rule,
    flat_graph_builder: &mut FlatGraphBuilder,
    tee_counter: &mut HashMap<String, Counter>,
    merge_counter: &mut HashMap<String, Counter>,
    next_join_idx: &mut Counter,
    diagnostics: &mut Vec<Diagnostic>,
) {
    let target = &rule.target.name;
    let target_ident = syn::Ident::new(&target.name, Span::call_site());

    let sources: Vec<Atom> = rule.sources.to_vec();

    // TODO(shadaj): smarter plans
    let mut plan: JoinPlan = sources
        .iter()
        .filter_map(|x| match x {
            Atom::Relation(negated, e) => {
                if negated.is_none() {
                    Some(JoinPlan::Source(e))
                } else {
                    None
                }
            }
            _ => None,
        })
        .reduce(|a, b| JoinPlan::Join(Box::new(a), Box::new(b)))
        .unwrap();

    plan = sources
        .iter()
        .filter_map(|x| match x {
            Atom::Relation(negated, e) => {
                if negated.is_some() {
                    Some(JoinPlan::Source(e))
                } else {
                    None
                }
            }
            _ => None,
        })
        .fold(plan, |a, b| JoinPlan::AntiJoin(Box::new(a), Box::new(b)));

    let predicates = sources
        .iter()
        .filter_map(|x| match x {
            Atom::Predicate(e) => Some(e),
            _ => None,
        })
        .collect::<Vec<_>>();

    if !predicates.is_empty() {
        plan = JoinPlan::Predicate(predicates, Box::new(plan))
    }

    let out_expanded = expand_join_plan(&plan, flat_graph_builder, tee_counter, next_join_idx);

    let after_join = apply_aggregations(rule, &out_expanded, diagnostics);

    let my_merge_index = merge_counter
        .entry(target.name.clone())
        .or_insert_with(|| 0..)
        .next()
        .expect("Out of merge indices");

    let my_merge_index_lit = syn::LitInt::new(&format!("{}", my_merge_index), Span::call_site());

    let after_join_and_send: Pipeline = match rule.rule_type {
        RuleType::Sync(_) => {
            if rule.target.at_node.is_some() {
                panic!("Rule must be async to send data to other nodes")
            }

            parse_quote!(#after_join -> [#my_merge_index_lit] #target_ident)
        }
        RuleType::NextTick(_) => {
            if rule.target.at_node.is_some() {
                panic!("Rule must be async to send data to other nodes")
            }

            parse_quote!(#after_join -> next_tick() -> [#my_merge_index_lit] #target_ident)
        }
        RuleType::Async(_) => {
            if rule.target.at_node.is_none() {
                panic!("Async rules are only for sending data to other nodes")
            }

            let exprs_get_data = rule
                .target
                .fields
                .iter()
                .enumerate()
                .map(|(i, _)| -> syn::Expr {
                    let syn_index = syn::Index::from(i);
                    parse_quote!(v.#syn_index)
                });

            let syn_target_index = syn::Index::from(rule.target.fields.len());

            let v_type = repeat_tuple::<syn::Type, syn::Type>(
                || parse_quote!(_),
                rule.target.fields.len() + 1,
            );

            let send_pipeline_ident = syn::Ident::new(
                &format!("{}_async_send", &rule.target.name.name),
                Span::call_site(),
            );

            parse_quote!(#after_join -> map(|v: #v_type| (v.#syn_target_index, (#(#exprs_get_data, )*))) -> #send_pipeline_ident)
        }
    };

    let out_name = out_expanded.name;
    // If the output comes with a tee index, we must read with that. This only happens when we are
    // directly outputting a transformation of a single relation on the RHS.
    let out_indexing = out_expanded.tee_idx.map(|i| Indexing {
        bracket_token: syn::token::Bracket::default(),
        index: hydroflow_lang::parse::PortIndex::Int(IndexInt {
            value: i,
            span: Span::call_site(),
        }),
    });
    flat_graph_builder.add_statement(hydroflow_lang::parse::HfStatement::Pipeline(
        Pipeline::Link(PipelineLink {
            lhs: Box::new(parse_quote!(#out_name #out_indexing)), // out_name[idx]
            arrow: parse_quote!(->),
            rhs: Box::new(after_join_and_send),
        }),
    ));
}

fn apply_aggregations(
    rule: &Rule,
    out_expanded: &IntermediateJoinNode,
    diagnostics: &mut Vec<Diagnostic>,
) -> Pipeline {
    let mut aggregations = vec![];
    let mut group_by_exprs = vec![];
    let mut agg_exprs = vec![];

    for field in rule
        .target
        .fields
        .iter()
        .chain(rule.target.at_node.iter().map(|n| &n.node))
    {
        let expr: syn::Expr = if let Some(col) = out_expanded
            .variable_mapping
            .get(&syn::Ident::new(&field.ident().name, Span::call_site()))
        {
            let source_col_idx = syn::Index::from(*col);
            parse_quote!(row.#source_col_idx)
        } else {
            diagnostics.push(Diagnostic::spanned(
                Span::call_site(),
                Level::Error,
                format!(
                    "Could not find column {} in RHS of rule",
                    &field.ident().name
                ),
            ));
            parse_quote!(())
        };

        match field {
            TargetExpr::Ident(_) => {
                group_by_exprs.push(expr);
            }
            TargetExpr::Aggregation(_) => {
                aggregations.push(field);
                agg_exprs.push(expr);
            }
        }
    }

    let flattened_tuple_type = &out_expanded.tuple_type;

    if agg_exprs.is_empty() {
        parse_quote!(map(|row: #flattened_tuple_type| (#(#group_by_exprs, )*)))
    } else {
        let agg_initial =
            repeat_tuple::<syn::Expr, syn::Expr>(|| parse_quote!(None), agg_exprs.len());

        let group_by_input_type =
            repeat_tuple::<syn::Type, syn::Type>(|| parse_quote!(_), group_by_exprs.len());

        let agg_input_type =
            repeat_tuple::<syn::Type, syn::Type>(|| parse_quote!(_), agg_exprs.len());
        let agg_type: syn::Type =
            repeat_tuple::<syn::Type, syn::Type>(|| parse_quote!(Option<_>), agg_exprs.len());

        let group_by_stmts: Vec<syn::Stmt> = aggregations
            .iter()
            .enumerate()
            .map(|(i, agg)| {
                let idx = syn::Index::from(i);
                let old_at_index: syn::Expr = parse_quote!(old.#idx);
                let val_at_index: syn::Expr = parse_quote!(val.#idx);

                let agg_expr: syn::Expr = match agg {
                    TargetExpr::Aggregation(Aggregation { tpe, .. }) => match tpe {
                        AggregationType::Max(_) => {
                            parse_quote!(std::cmp::max(prev, #val_at_index))
                        }
                    },
                    _ => panic!(),
                };

                parse_quote! {
                    #old_at_index = if let Some(prev) = #old_at_index {
                        Some(#agg_expr)
                    } else {
                        Some(#val_at_index)
                    };
                }
            })
            .collect();

        let group_by_fn: syn::Expr = parse_quote!(|old: &mut #agg_type, val: #agg_input_type| {
            #(#group_by_stmts)*
        });

        let mut after_group_lookups: Vec<syn::Expr> = vec![];
        let mut group_key_idx = 0;
        let mut agg_idx = 0;
        for field in rule
            .target
            .fields
            .iter()
            .chain(rule.target.at_node.iter().map(|n| &n.node))
        {
            match field {
                TargetExpr::Ident(_) => {
                    let idx = syn::Index::from(group_key_idx);
                    after_group_lookups.push(parse_quote!(g.#idx));
                    group_key_idx += 1;
                }
                TargetExpr::Aggregation(_) => {
                    let idx = syn::Index::from(agg_idx);
                    after_group_lookups.push(parse_quote!(a.#idx.unwrap()));
                    agg_idx += 1;
                }
            }
        }

        let pre_group_by_map: syn::Expr = parse_quote!(|row: #flattened_tuple_type| ((#(#group_by_exprs, )*), (#(#agg_exprs, )*)));
        let after_group_map: syn::Expr = parse_quote!(|(g, a)| (#(#after_group_lookups, )*));

        parse_quote! {
            map(#pre_group_by_map) -> group_by::<'tick, #group_by_input_type, #agg_type>(|| #agg_initial, #group_by_fn) -> map(#after_group_map)
        }
    }
}

#[cfg(test)]
mod tests {
    use syn::parse_quote;

    use super::gen_hydroflow_graph;
    use super::hydroflow_graph_to_program;

    macro_rules! test_snapshots {
        ($program:literal) => {
            let flat_graph = gen_hydroflow_graph(parse_quote!($program)).unwrap();

            let flat_graph_ref = &flat_graph;
            insta::with_settings!({snapshot_suffix => "surface_graph"}, {
                insta::assert_display_snapshot!(flat_graph_ref.surface_syntax_string());
            });

            let out = &hydroflow_graph_to_program(flat_graph, quote::quote! { hydroflow });
            let wrapped: syn::File = parse_quote! {
                fn main() {
                    #out
                }
            };

            insta::with_settings!({snapshot_suffix => "datalog_program"}, {
                insta::assert_display_snapshot!(
                    prettyplease::unparse(&wrapped)
                );
            });
        };
    }

    #[test]
    fn minimal_program() {
        test_snapshots!(
            r#"
            .input input `source_stream(input)`
            .output out `for_each(|v| out.send(v).unwrap())`

            out(y, x) :- input(x, y).
            "#
        );
    }

    #[test]
    fn join_with_self() {
        test_snapshots!(
            r#"
            .input input `source_stream(input)`
            .output out `for_each(|v| out.send(v).unwrap())`

            out(x, y) :- input(x, y), input(y, x).
            "#
        );
    }

    #[test]
    fn join_with_other() {
        test_snapshots!(
            r#"
            .input in1 `source_stream(in1)`
            .input in2 `source_stream(in2)`
            .output out `for_each(|v| out.send(v).unwrap())`

            out(x, y) :- in1(x, y), in2(y, x).
            "#
        );
    }

    #[test]
    fn multiple_contributors() {
        test_snapshots!(
            r#"
            .input in1 `source_stream(in1)`
            .input in2 `source_stream(in2)`
            .output out `for_each(|v| out.send(v).unwrap())`

            out(x, y) :- in1(x, y).
            out(x, y) :- in2(y, x).
            "#
        );
    }

    #[test]
    fn transitive_closure() {
        test_snapshots!(
            r#"
            .input edges `source_stream(edges)`
            .input seed_reachable `source_stream(seed_reachable)`
            .output reachable `for_each(|v| reachable.send(v).unwrap())`

            reachable(x) :- seed_reachable(x).
            reachable(y) :- reachable(x), edges(x, y).
            "#
        );
    }

    #[test]
    fn single_column_program() {
        test_snapshots!(
            r#"
            .input in1 `source_stream(in1)`
            .input in2 `source_stream(in2)`
            .output out `for_each(|v| out.send(v).unwrap())`

            out(x) :- in1(x), in2(x).
            "#
        );
    }

    #[test]
    fn triple_relation_join() {
        test_snapshots!(
            r#"
            .input in1 `source_stream(in1)`
            .input in2 `source_stream(in2)`
            .input in3 `source_stream(in3)`
            .output out `for_each(|v| out.send(v).unwrap())`

            out(d, c, b, a) :- in1(a, b), in2(b, c), in3(c, d).
            "#
        );
    }

    #[test]
    fn local_constraints() {
        test_snapshots!(
            r#"
            .input input `source_stream(input)`
            .output out `for_each(|v| out.send(v).unwrap())`

            out(x, x) :- input(x, x).
            "#
        );

        test_snapshots!(
            r#"
            .input input `source_stream(input)`
            .output out `for_each(|v| out.send(v).unwrap())`

            out(x, x, y, y) :- input(x, x, y, y).
            "#
        );
    }

    #[test]
    fn test_simple_filter() {
        test_snapshots!(
            r#"
            .input input `source_stream(input)`
            .output out `for_each(|v| out.send(v).unwrap())`

            out(x, y) :- input(x, y), ( x > y ), ( y == x ).
            "#
        );
    }

    #[test]
    fn test_anti_join() {
        test_snapshots!(
            r#"
            .input ints_1 `source_stream(ints_1)`
            .input ints_2 `source_stream(ints_2)`
            .input ints_3 `source_stream(ints_3)`
            .output result `for_each(|v| result.send(v).unwrap())`

            result(x, z) :- ints_1(x, y), ints_2(y, z), !ints_3(y)
            "#
        );
    }

    #[test]
    fn test_max() {
        test_snapshots!(
            r#"
            .input ints `source_stream(ints)`
            .output result `for_each(|v| result.send(v).unwrap())`

            result(max(a), b) :- ints(a, b)
            "#
        );
    }

    #[test]
    fn test_max_all() {
        test_snapshots!(
            r#"
            .input ints `source_stream(ints)`
            .output result `for_each(|v| result.send(v).unwrap())`

            result(max(a), max(b)) :- ints(a, b)
            "#
        );
    }

    #[test]
    fn test_send_to_node() {
        test_snapshots!(
            r#"
            .input ints `source_stream(ints)`
            .output result `for_each(|v| result.send(v).unwrap())`
            .async result `for_each(|(node, data)| async_send_result(node, data))` `source_stream(async_receive_result)`

            result@b(a) :~ ints(a, b)
            "#
        );
    }
}
