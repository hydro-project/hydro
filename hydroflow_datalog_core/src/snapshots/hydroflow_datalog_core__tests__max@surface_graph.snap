---
source: hydroflow_datalog_core/src/lib.rs
expression: graph.surface_syntax_string()
---
1v1 = merge ();
2v1 = unique :: < 'tick > ();
3v1 = tee ();
4v1 = merge ();
5v1 = unique :: < 'tick > ();
6v1 = tee ();
7v1 = source_stream (ints);
8v1 = for_each (| v | result . send (v) . unwrap ());
9v1 = map (| row : (_ , _ ,) | ((row . 1 ,) , (row . 0 ,)));
10v1 = group_by :: < 'tick , (_ ,) , (Option < _ > ,) > (| | (None ,) , | old : & mut (Option < _ > ,) , val : (_ ,) | { old . 0 = if let Some (prev) = old . 0 { Some (std :: cmp :: max (prev , val . 0)) } else { Some (val . 0) } ; });
11v1 = map (| (g , a) | (a . 0 . unwrap () , g . 0 ,));

2v1 -> 3v1;
1v1 -> 2v1;
5v1 -> 6v1;
4v1 -> 5v1;
7v1 -> 1v1;
6v1 -> 8v1;
11v1 -> 4v1;
10v1 -> 11v1;
9v1 -> 10v1;
3v1 -> 9v1;

